src/components/fields/MarkerColor.js

```diff
@@ -0,0 +1,142 @@
+import Field from './Field';
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import {connectToContainer} from 'lib';
+import RadioBlocks from '../widgets/RadioBlocks';
+import Color from './Color';
+import DataSelector from './DataSelector';
+import Colorscale from './Colorscale';
+import {MULTI_VALUED, COLORS} from 'lib/constants';
+
+class UnconnectedMarkerColor extends Component {
+  constructor(props, context) {
+    super(props, context);
+
+    let type = null;
+    if (
+      !props.container.marker ||
+      (props.container.marker && !props.container.marker.colorsrc)
+    ) {
+      type = 'constant';
+    } else if (
+      props.container.marker &&
+      Array.isArray(props.container.marker.color) &&
+      props.fullContainer.marker &&
+      Array.isArray(props.fullContainer.marker.color)
+    ) {
+      type = 'variable';
+    }
+
+    this.state = {
+      type,
+      value: {
+        constant: type === 'constant' ? props.fullValue : COLORS.mutedBlue,
+        variable: type === 'variable' ? props.fullValue : null,
+      },
+    };
+
+    this.setType = this.setType.bind(this);
+    this.setValue = this.setValue.bind(this);
+    this.setColorScale = this.setColorScale.bind(this);
+  }
+
+  setType(type) {
+    this.setState({type: type});
+    this.props.updatePlot(this.state.value[type]);
+    if (type === 'constant') {
+      this.context.updateContainer({
+        ['marker.colorsrc']: null,
+        ['marker.colorscale']: null,
+      });
+      this.setState({colorscale: null});
+    } else {
+      this.context.updateContainer({
+        ['marker.color']: null,
+        ['marker.colorsrc']: null,
+        ['marker.colorscale']: [],
+      });
+    }
+  }
+
+  setValue(inputValue) {
+    const {type} = this.state;
+
+    this.setState(
+      type === 'constant'
+        ? {value: {constant: inputValue}}
+        : {value: {variable: inputValue}}
+    );
+    this.props.updatePlot(inputValue);
+  }
+
+  setColorScale(inputValue) {
+    this.setState({colorscale: inputValue});
+    this.context.updateContainer({['marker.colorscale']: inputValue});
```

Comment:
we don't have to update colorsrc?

Reply:
Why would we? We're just changing colorscale. Am I missing something?

Reply:
nvm, forgot how it works